packages/malloc/src/pool.ts

```diff
@@ -10,42 +10,108 @@ import {
 } from "./api";
 import { wrap } from "./wrap";
 
+const POINTER_TOP_OFFSET = 0;
+const POINTER_END_OFFSET = POINTER_TOP_OFFSET + Uint32Array.BYTES_PER_ELEMENT;
+const POINTER_FREE_OFFSET = POINTER_END_OFFSET + Uint32Array.BYTES_PER_ELEMENT;
+const POINTER_USED_OFFSET = POINTER_FREE_OFFSET + Uint32Array.BYTES_PER_ELEMENT;
+
+const DATA_SPACE_OFFSET = POINTER_USED_OFFSET + Uint32Array.BYTES_PER_ELEMENT;
+
+
 export class MemPool implements IMemPool {
     buf: ArrayBuffer;
-    protected top: number;
-    protected start: number;
-    protected end: number;
+
+    /**
+     * Start is the anchor index inside the arraybuffer, so we can't save it inside the arraybuffer itself.
+     * If you pass the ArrayBuffer to other consumers they must use the same start value
+    */
+    protected readonly start: number;
+
     protected doCompact: boolean;
     protected doSplit: boolean;
     protected minSplit: number;
-    protected _free: MemBlock | null;
-    protected _used: MemBlock | null;
-
     protected u8: Uint8Array;
+    protected dataView: DataView;
 
     constructor(opts: Partial<MemPoolOpts> = {}) {
         this.buf = opts.buf ? opts.buf : new ArrayBuffer(opts.size || 0x1000);
         this.u8 = new Uint8Array(this.buf);
-        this.start = opts.start != null ? align(Math.max(opts.start, 8), 8) : 8;
-        this.end =
-            opts.end != null
-                ? Math.min(opts.end, this.buf.byteLength)
-                : this.buf.byteLength;
-        if (this.start >= this.end) {
-            illegalArgs(
-                `invalid address range (0x${this.start.toString(
-                    16
-                )} - 0x${this.end.toString(16)})`
-            );
-        }
-        this.top = this.start;
+        this.dataView = new DataView(this.buf);
```

Comment:
Using using a `Uint32Array` for the state information would simplify those other state accessors further down:

```ts
this.state = new Uint32Array(this.buf, this.start, NUM_STATE_FIELDS);
```

(then also should move it to after `this.start = ...` 😄)